sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -30,6 +30,13 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.types.UTF8String
 
+/**
+ * This trait is to indicate that this is an expression that creates a collection
+ * that will not be null and will not be empty when it contains children.
+ * Note that it will be foldable if it doesn't container children.
+ */
+trait CreateNonEmptyNonNullCollection
```

Comment:
Maybe `CreateNonNullCollection`? The complex type creators `CreateArray` and `CreateMap` creates non null collection, but I think we cannot infer it creates non empty collection. Although the comment said it is only non empty if the creator contains children, it cannot be inferred from the trait itself only.